test/e2e/lib/jest/environment.js

```diff
@@ -19,6 +19,9 @@ class JestEnvironmentE2E extends JestEnvironmentNode {
 				break;
 
 			case 'hook_failure':
+				console.error( event.error );
```

Comment:
We don't want to log the error in case of `test_fn_failure`? Does jest does it automatically?

At the very least I think we should set `this.global.__CURRENT_TEST_FAILED__ = true;` here. That global property is checked by the screenshot/video recorder hooks.

Most of the time a failed hook mean there is nothing to take a screenshot from, but we can't guarantee that will be the case forever (for example, we may have a hook that opens the browser and does some navigation, and we do want to screenshot that if it fails).

Reply:
I think my comment also went missing too 👀 

> We don't want to log the error in case of test_fn_failure? Does jest does it automatically?

In our current configuration, Jest automatically throws and outputs the relevant logs if the error is thrown outside of a hook.

>  At the very least I think we should set this.global.__CURRENT_TEST_FAILED__ = true; here. That global property is checked by the screenshot/video recorder hooks.

Avoiding the screenshot/video recording codepath was why initially the value was not set. But as you mention:

> we can't guarantee that will be the case forever (for example, we may have a hook that opens the browser and does some navigation, and we do want to screenshot that if it fails)

Yes, that is true. I will treat hook failure like a normal test failure in next commit.

Reply:
Fixed in 0581a7d363a18ef36f40761bcc6e92aae9f5e83d.

Revised behavior now outputs the following if anything in the hook fails:

```
 FAIL  specs/specs-playwright/wp-media__edit-spec.ts
  ● Test suite failed to run

    Source file /Users/edwintakahashi/workspace/wp-calypso-second/test/e2e/image-uploads/image0.jpg not found on disk.

      78 | 		await fs.access( 'sourcePath' );
      79 | 	} catch {
    > 80 | 		throw new Error( `Source file ${ sourcePath } not found on disk.` );
         | 		      ^
      81 | 	}
      82 |
      83 | 	// Obtain the file extension.

      at Object.createTestFile (../../packages/calypso-e2e/src/media-helper.ts:80:9)
      at specs/specs-playwright/wp-media__edit-spec.ts:22:15

Test Suites: 1 failed, 1 total
Tests:       0 total
Snapshots:   0 total
Time:        2.705 s
```

Reply:
I've revised the behavior again in 87006bedbe2d81a6b079cd85db0bf4853bee05ed since I felt the previous behavior was a bit hacky. 

Recap: throwing an error in `environment.js` causes rest of execution to halt. Practically, this means the `afterAll` hooks of screenshot/video capture are not run.

I've changed the behavior to instead do the following:

1. if a hook failure is detected, set a new attribute `hookFailed`.
2. at `test_start` event, check if hook has failed.
3. if hook has failed, mark all test cases in the describe block as failed.

I arrived at this solution because [these lines](https://github.com/Automattic/wp-calypso/blob/trunk/test/e2e/lib/jest/environment.js#L16-L18) were causing hook failures to be treated like a normal skip and ended up suppressing the errors thrown. To verify this, add an exception in a hook, modify the `skip` to a `fail`. Errors thrown in the hook should now be visible. 

The separate treatment of hook failure is required, otherwise if an exception is thrown in the test step it would not be skipped. This can be verified by performing the same steps as above, but place the exception in the test step instead. Jest will not skip rest of the test steps.

Try pulling my changes in this revision, place exceptions in beforeAll hook, test step and afterAll hook. It should behave following this rule:
1. if exception is thrown in a hook, suite is marked as failed and test steps yet to run are also marked as failed.
2. if exception is thrown in a step, only the failing step is marked failed and rest are skipped.

Reply:
Scenario 1:
```
  ● [WPCOM] Media: Edit Media: (desktop) @parallel › Edit Image (Atomic) › Cancel image edit

    hello!

      20 |
      21 | 	beforeAll( async () => {
    > 22 | 		throw new Error( 'hello!' );
         | 		      ^
      23 | 		testImage = await MediaHelper.createTestFile( TEST_IMAGE_PATH );
      24 | 	} );
      25 |

      at specs/specs-playwright/wp-media__edit-spec.ts:22:9

Test Suites: 1 failed, 1 total
Tests:       16 failed, 16 total
Snapshots:   0 total
Time:        3.14 s, estimated 12 s
```

Scenario 2:
```
  ● [WPCOM] Media: Edit Media: (desktop) @parallel › Edit Image (Simple) › Log In

    hello!

      31 |
      32 | 		it( 'Log In', async function () {
    > 33 | 			throw new Error( 'hello!' );
         | 			      ^
      34 | 			const loginFlow = new LoginFlow( page, user );
      35 | 			await loginFlow.logIn();
      36 | 		} );

      at Object.<anonymous> (specs/specs-playwright/wp-media__edit-spec.ts:33:10)

Test Suites: 1 failed, 1 total
Tests:       1 failed, 15 skipped, 16 total
Snapshots:   0 total
Time:        3.078 s, estimated 4 s
```

Scenario 3:
```
  ● Test suite failed to run

    hello!

      60 |
      61 | 		afterAll( async function () {
    > 62 | 			throw new Error( 'hello!' );
         | 			      ^
      63 | 		} );
      64 | 	} );
      65 | } );

      at specs/specs-playwright/wp-media__edit-spec.ts:62:10
          at runMicrotasks (<anonymous>)

Test Suites: 1 failed, 1 total
Tests:       2 passed, 2 total
Snapshots:   0 total
Time:        11.912 s, estimated 43 s
```

Reply:
@scinos Could I please get a review on this before it gets too stale? Thanks!